### PR TITLE
[SPARK-15357] Cooperative spilling should check consumer memory mode

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

Since we support forced spilling for Spillable, which only works in OnHeap mode, different from other SQL operators (could be OnHeap or OffHeap), we should considering the mode of consumer before calling trigger forced spilling.
## How was this patch tested?

Add new test.

### CODE REVIEW

**Comment by davies** [comment]
cc @andrewor14

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 1]
```diff
@@ -149,10 +145,10 @@ public long acquireExecutionMemory(
       if (got < required) {
         // Call spill() on other consumers to release memory
         for (MemoryConsumer c: consumers) {
-          if (c != consumer && c.getUsed() > 0) {
+          if (c != consumer && c.getUsed() > 0 && c.getMode() == mode) {
```
Comment (andrewor14) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java]: great

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -281,20 +281,20 @@ private[spark] class Executor(
           val releasedLocks = env.blockManager.releaseAllLocksForTask(taskId)
           val freedMemory = taskMemoryManager.cleanUpAllAllocatedMemory()
 
-          if (freedMemory > 0) {
+          if (freedMemory > 0 && !threwException) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: this is not related right? Just wondering

Comment (davies) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Not, it's an minor thing maybe not worth an PR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 3]
```diff
@@ -398,25 +398,21 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
       0)
   }
 
-  private def acquireMemory(size: Long): Unit = {
+  private def ensureAcquireMemory(size: Long): Unit = {
     // do not support spilling
-    val got = mm.acquireExecutionMemory(size, MemoryMode.ON_HEAP, this)
+    val got = acquireMemory(size)
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: where is this defined?

**Comment by andrewor14** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: yeah, looks good

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 3 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: This is defined in MemoryConsumer

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!
